### PR TITLE
Implement git refs API for listing references (branches, tags and other)

### LINKED ISSUE #2738: API support for listing tags and fetching tag details
## Description

The API currently exposes details of branches and pull requests but does not allow discovering tags.

The minimum support that I would like to see:

* A way to get a list of all tags in a repository, e.g. `/api/v1/repos/${repoOwner}/${repoName}/tags`
* A way to get a specific tag, e.g. `/api/v1/repos/${repoOwner}/${repoName}/tags/${tagName}`
* Push event notification on creation of tags (if not already present)

The tag JSON object should probably have the following structure:

* lightweight tags (i.e. where the tag refers to a commit):
  ```
  {
    name: "v1.0.0",
    commit: {
      id: "...",
      message: "...",
      url: "...",
      author: {
        name: "...",
        email: "...",
        username: "..."
      },
      committer: {
        name: "...",
        email: "...",
        username: "..."
      },
      verification: {
        verified: false,
        reason: "gpg.error.not_signed_commit",
        signature: "",
        payload: ""
      },
      timestamp: "2017-08-31T19:26:04Z"
    }
  }
  ```

* annotated tags (i.e. where the tag is an object referencing the commit):
  ```
  {
    name: "v1.0.0",
    tag: {
      id: "...",
      message: "...",
      url: "...",
      author: {
        name: "...",
        email: "...",
        username: "..."
      },
      committer: {
        name: "...",
        email: "...",
        username: "..."
      },
      verification: {
        verified: false,
        reason: "gpg.error.not_signed_commit",
        signature: "",
        payload: ""
      },
      timestamp: "2017-08-31T19:26:04Z",
      commit: {
        id: "...",
        message: "...",
        url: "...",
        author: {
          name: "...",
          email: "...",
          username: "..."
        },
        committer: {
          name: "...",
          email: "...",
          username: "..."
        },
        verification: {
          verified: false,
          reason: "gpg.error.not_signed_commit",
          signature: "",
          payload: ""
        },
        timestamp: "2017-08-31T19:26:04Z"
      }
    }
  }
  ```

### PR DESCRIPTION
**Pull request by lafriks**
Implements git refs API `/repos/{owner}/{repo}/git/refs/{ref}`. Implements only `GET` methods fully compatible with GitHub API

Partly fixes  #2738

Depends on go-gitea/go-sdk#129

TODO:
- [x] SDK pull request
- [x] Tests

### CODE REVIEW

File: routers/api/v1/repo/git_ref.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
+// Copyright 2018 Gitea. All rights reserved.
```
Comment (lunny) [inline routers/api/v1/repo/git_ref.go]: The Gitea Authors

File: routers/api/v1/repo/git_ref.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+package repo
+
+import (
+	"code.gitea.io/gitea/modules/context"
+
+	"code.gitea.io/git"
+	api "code.gitea.io/sdk/gitea"
+)
+
+// GetGitRefs get ref or an list all the refs of a repository
```
Comment (adelowo) [inline routers/api/v1/repo/git_ref.go]:
```suggestion
// GetGitRefs gets the ref or a list of all the refs of a repository
```

**Comment by techknowlogick** [comment]
@lafriks please update vendored files, and resolve conflict.

**Comment by lafriks** [comment]
Ready for review

File: routers/api/v1/repo/git_ref.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	apiRefs := make([]*api.Reference, len(refs))
+	for i := range refs {
+		apiRefs[i] = &api.Reference{
+			Ref: refs[i].Name,
+			URL: ctx.Repo.Repository.APIURL() + "/git/" + refs[i].Name,
+			Object: &api.GitObject{
+				SHA:  refs[i].Object.String(),
+				Type: refs[i].Type,
+				// TODO: Add commit/tag info URL
+				//URL:  ctx.Repo.Repository.APIURL() + "/git/" + refs[i].Type + "s/" + refs[i].Object.String(),
```
Comment (lunny) [inline routers/api/v1/repo/git_ref.go:105]: Not finished?

File: routers/api/v1/api.go [hunk 4]
```diff
@@ -573,6 +573,10 @@ func RegisterRoutes(m *macaron.Macaron) {
 					m.Get("/status", repo.GetCombinedCommitStatusByRef)
 					m.Get("/statuses", repo.GetCommitStatusesByRef)
 				})
+				m.Group("/git", func() {
```
Comment (lunny) [inline routers/api/v1/api.go:576]: permission check?

**Review by lunny — CHANGES_REQUESTED** [review]

File: routers/api/v1/repo/git_ref.go (hunk 3 above)
Comment (lafriks) [reply routers/api/v1/repo/git_ref.go:105]: That is for next pull request where I will implement getting tags details in route `/git/tags`

File: routers/api/v1/api.go (hunk 4 above)
Comment (lafriks) [reply routers/api/v1/api.go:576]: what exact permission check are you talking about. (It is checked for access read only rights for repo). Do you mean unit type check?

Comment (lunny) [reply routers/api/v1/api.go:576]: Didn't find the access read check. Unit type check should be fixed by my PR #5314 .

Comment (lafriks) [reply routers/api/v1/api.go:576]: It is already checked in upper group in function `repoAssignment`

**Comment by lafriks** [comment]
@lunny ? :)

File: routers/api/v1/api.go (hunk 4 above)
Comment (lunny) [reply routers/api/v1/api.go:576]: Right. It's different from the middleware which has the same name and used by ui routers.

**Comment by lafriks** [comment]
@lunny so is there anything I need to fix this PR?

**Review by lunny — APPROVED** [review]

**Review by techknowlogick — APPROVED** [review]

**Pull request merged**